spring-integration-core/src/main/java/org/springframework/integration/channel/management/DefaultMessageChannelMetrics.java

```diff
@@ -77,118 +77,178 @@ public void destroy() {
 		}
 	}
 
-	public String getName() {
-		return name;
-	}
-
-	public long beforeSend() {
+	@Override
+	public MetricsContext beforeSend() {
 		if (logger.isTraceEnabled()) {
 			logger.trace("Recording send on channel(" + this.name + ")");
 		}
 
 		long start = 0;
 		if (this.fullStatsEnabled) {
-			start = System.currentTimeMillis();
+			start = System.nanoTime();
 			this.sendRate.increment();
 		}
 		this.sendCount.incrementAndGet();
 
-		return start;
+		return new DefaultChannelMetricsContext(start);
 	}
 
-	public void afterSend(long start, boolean result) {
-		if (start > 0) {
-			long now = System.currentTimeMillis();
-			long elapsed = now - start;
-			if (result && this.fullStatsEnabled) {
-				sendSuccessRatio.success(now);
-				sendDuration.append(elapsed);
-			}
-			else {
-				if (this.fullStatsEnabled) {
-	 				sendSuccessRatio.failure(now);
-					sendErrorCount.incrementAndGet();
-				}
-				sendErrorRate.increment();
-			}
-			if (logger.isTraceEnabled()) {
-				logger.trace("Elapsed: " + this.name + ": " + elapsed);
+	@Override
+	public void afterSend(MetricsContext context, boolean result) {
+		long start = ((DefaultChannelMetricsContext) context).start;
```

Comment:
I made sure it couldn't happen in the calling code; since we are trying to streamline performance I didn't want to add the overhead of a null check when it's unnecessary but I guess it's not a big overhead.